### ISSUE TITLE
📝 Formatter still not idempotent on member chains with object-literal args (2.4.16) — same as closed #4383

### ISSUE THREAD

**Issue by vzd3v** [biomejs/biome#10531]
### Summary

The JS formatter is **not idempotent** for a member/method chain whose final call argument is an object literal and whose single-line form exceeds the print width. The first `format` pass fully expands the chain (one call per line); the second pass collapses the chain back inline and breaks only the object argument; the third pass is stable.

Because of this, a single `biome format --write` (or `biome check --write`) leaves the file in the *first-pass* shape, which a subsequent `biome format` / `biome check` then rejects. In CI this manifests as `biome check` failing on code that was just formatted.

This is the exact behavior reported in #4383, which was closed on 2026-03-02 as "seems fixed". It still reproduces on the **latest release, 2.4.16** — including the original `drizzle-kit` snippet from that issue. Filing a fresh report rather than commenting on the closed one; please re-link/close as duplicate if preferred.

### Environment

```
Version: 2.4.16   (npm dist-tag: latest)
```

`biome.json`:

```json
{
  "formatter": { "indentStyle": "tab", "lineWidth": 120 }
}
```

### Steps to reproduce

`repro.js` (the original snippet from #4383):

```js
import { integer, pgTable } from "drizzle-orm/pg-core";

const example = pgTable("example", {
	id: integer().primaryKey().generatedByDefaultAsIdentity({ name: "example_id_seq", startWith: 1, increment: 1, minValue: 1, maxValue: 2147483647, cache: 1 }),
});
```

```
biome format --write repro.js   # pass 1
biome format --write repro.js   # pass 2  -> different output
```

**After pass 1** (chain expanded):

```js
const example = pgTable("example", {
	id: integer()
		.primaryKey()
		.generatedByDefaultAsIdentity({
			name: "example_id_seq",
			startWith: 1,
			increment: 1,
			minValue: 1,
			maxValue: 2147483647,
			cache: 1,
		}),
});
```

**After pass 2** (chain collapsed inline — this is the stable fixed point; pass 3 == pass 2):

```js
const example = pgTable("example", {
	id: integer().primaryKey().generatedByDefaultAsIdentity({
		name: "example_id_seq",
		startWith: 1,
		increment: 1,
		minValue: 1,
		maxValue: 2147483647,
		cache: 1,
	}),
});
```

### Not specific to drizzle

The trigger is purely the chain shape, not any particular API. A plain Fastify-style handler reproduces it identically (`lineWidth: 100`, `indentStyle: space`):

```js
function handler(reply) {
	return reply.code(409).send({ ok: false, error: 'User with this email already exists here', code: 'DUPLICATE_EMAIL' });
}
```

- pass 1 → `return reply\n  .code(409)\n  .send({ … })` (expanded)
- pass 2 → `return reply.code(409).send({ … })` (collapsed, stable)

ASCII vs non-ASCII string contents make no difference, so this is not a Unicode width-measurement issue — it is the member-chain break decision.

### Expected behavior

`format(format(x)) === format(x)` — a single pass should already produce the stable (pass-2) shape. Currently two passes are required to converge.

### Actual behavior

Pass 1 ≠ pass 2; pass 2 == pass 3. One `format --write` is insufficient, and a follow-up `biome check` fails on freshly-formatted code.

### Notes

- Reproduces on `latest` (2.4.16).
- Appears to be the same root cause as #4383 (closed as "seems fixed").

**Comment by dyc3** [comment]
playground to reproduce: https://biomejs.dev/playground/?lineWidth=120&tab=formatter&pane=Diagnostics&code=aQBtAHAAbwByAHQAIAB7AGkAbgB0AGUAZwBlAHIALAAgAHAAZwBUAGEAYgBsAGUAfQAgAGYAcgBvAG0AIAAiAGQAcgBpAHoAegBsAGUALQBvAHIAbQAvAHAAZwAtAGMAbwByAGUAIgA7AAoACgBjAG8AbgBzAHQAIABlAHgAYQBtAHAAbABlACAAPQAgAHAAZwBUAGEAYgBsAGUAKAAiAGUAeABhAG0AcABsAGUAIgAsACAAewAKAAkAaQBkADoAIABpAG4AdABlAGcAZQByACgAKQAuAHAAcgBpAG0AYQByAHkASwBlAHkAKAApAC4AZwBlAG4AZQByAGEAdABlAGQAQgB5AEQAZQBmAGEAdQBsAHQAQQBzAEkAZABlAG4AdABpAHQAeQAoAHsAIABuAGEAbQBlADoAIAAiAGUAeABhAG0AcABsAGUAXwBpAGQAXwBzAGUAcQAiACwAIABzAHQAYQByAHQAVwBpAHQAaAA6ACAAMQAsACAAaQBuAGMAcgBlAG0AZQBuAHQAOgAgADEALAAgAG0AaQBuAFYAYQBsAHUAZQA6ACAAMQAsACAAbQBhAHgAVgBhAGwAdQBlADoAIAAyADEANAA3ADQAOAAzADYANAA3ACwAIABjAGEAYwBoAGUAOgAgADEAIAB9ACkACgB9ACkAOwAKAAoALwAvACAAYQBmAHQAZQByACAAZgBpAHIAcwB0ACAAZgBvAHIAbQBhAHQACgBpAG0AcABvAHIAdAAgAHsAIABpAG4AdABlAGcAZQByACwAIABwAGcAVABhAGIAbABlACAAfQAgAGYAcgBvAG0AIAAiAGQAcgBpAHoAegBsAGUALQBvAHIAbQAvAHAAZwAtAGMAbwByAGUAIgA7AAoACgBjAG8AbgBzAHQAIABlAHgAYQBtAHAAbABlACAAPQAgAHAAZwBUAGEAYgBsAGUAKAAiAGUAeABhAG0AcABsAGUAIgAsACAAewAKAAkAaQBkADoAIABpAG4AdABlAGcAZQByACgAKQAKAAkACQAuAHAAcgBpAG0AYQByAHkASwBlAHkAKAApAAoACQAJAC4AZwBlAG4AZQByAGEAdABlAGQAQgB5AEQAZQBmAGEAdQBsAHQAQQBzAEkAZABlAG4AdABpAHQAeQAoAHsACgAJAAkACQBuAGEAbQBlADoAIAAiAGUAeABhAG0AcABsAGUAXwBpAGQAXwBzAGUAcQAiACwACgAJAAkACQBzAHQAYQByAHQAVwBpAHQAaAA6ACAAMQAsAAoACQAJAAkAaQBuAGMAcgBlAG0AZQBuAHQAOgAgADEALAAKAAkACQAJAG0AaQBuAFYAYQBsAHUAZQA6ACAAMQAsAAoACQAJAAkAbQBhAHgAVgBhAGwAdQBlADoAIAAyADEANAA3ADQAOAAzADYANAA3ACwACgAJAAkACQBjAGEAYwBoAGUAOgAgADEALAAKAAkACQB9ACkALAAKAH0AKQA7AAoA

**Comment by dadavidtseng** [comment]
@dyc3 Hi, I'm a student in the CodePath AI301 program and I'd like to work on this issue. Could you assign it to me? I'll start by reproducing the problem locally and report here as I make progress. Any pointers on where to start in the codebase would be appreciated. Thank you!

**Comment by dyc3** [comment]
@dadavidtseng Go for it 👍
